hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/upgrade/TestDatanodeUpgradeToSchemaV3.java

```diff
@@ -0,0 +1,787 @@
+      ContainerProtos.ContainerCommandRequestProto request) {
+    dispatchRequest(request, ContainerProtos.Result.SUCCESS);
+  }
+
+  public void dispatchRequest(
+      ContainerProtos.ContainerCommandRequestProto request,
+      ContainerProtos.Result expectedResult) {
+    ContainerProtos.ContainerCommandResponseProto response =
+        dsm.getContainer().getDispatcher().dispatch(request, null);
+    Assert.assertEquals(expectedResult, response.getResult());
+  }
+
+  /// VOLUME OPERATIONS ///
+
+  /**
+   * Append a datanode volume to the existing volumes in the configuration.
+   * @return The root directory for the new volume.
+   */
+  public File addHddsVolume() throws Exception {
+    File vol = tempFolder.newFolder(UUID.randomUUID().toString());
+    String[] existingVolumes =
+        conf.getStrings(ScmConfigKeys.HDDS_DATANODE_DIR_KEY);
+    List<String> allVolumes = new ArrayList<>();
+    if (existingVolumes != null) {
+      allVolumes.addAll(Arrays.asList(existingVolumes));
+    }
+
+    allVolumes.add(vol.getAbsolutePath());
+    conf.setStrings(ScmConfigKeys.HDDS_DATANODE_DIR_KEY,
+        allVolumes.toArray(new String[0]));
+
+    return vol;
+  }
+
+  /**
+   * Append a db volume to the existing volumes in the configuration.
+   * @return The root directory for the new volume.
+   */
+  public File addDbVolume() throws Exception {
+    File vol = tempFolder.newFolder(UUID.randomUUID().toString());
+    String[] existingVolumes =
+        conf.getStrings(OzoneConfigKeys.HDDS_DATANODE_CONTAINER_DB_DIR);
+    List<String> allVolumes = new ArrayList<>();
+    if (existingVolumes != null) {
+      allVolumes.addAll(Arrays.asList(existingVolumes));
+    }
+
+    allVolumes.add(vol.getAbsolutePath());
+    conf.setStrings(OzoneConfigKeys.HDDS_DATANODE_CONTAINER_DB_DIR,
+        allVolumes.toArray(new String[0]));
+
+    return vol;
+  }
+
+  /**
+   * Renames the specified volume directory so it will appear as failed to
+   * the datanode.
+   */
+  public void failVolume(File volume) {
+    File failedVolume = getFailedVolume(volume);
+    Assert.assertTrue(volume.renameTo(failedVolume));
+  }
+
+  /**
+   * Convert the specified volume from its failed name back to its original
+   * name. The File passed should be the original volume path, not the one it
+   * was renamed to to fail it.
+   */
+  public void restoreVolume(File volume) {
+    File failedVolume = getFailedVolume(volume);
+    Assert.assertTrue(failedVolume.renameTo(volume));
+  }
+
+  /**
+   * @return The file name that will be used to rename a volume to fail it.
+   */
+  public File getFailedVolume(File volume) {
+    return new File(volume.getParent(), volume.getName() + "-failed");
+  }
+
+  /**
+   * Checks whether the datanode thinks the volume has failed.
+   * This could be outdated information if the volume was restored already
+   * and the datanode has not been restarted since then.
+   */
+  public boolean dnThinksVolumeFailed(File volume) {
```

Comment:
Ah, I happen to notice that there are some unused functions in this file from my Intellij (maybe copied from TestDatanodeUpgradeToScmHA.java), maybe we should remove them?